pulp_container/tests/functional/api/test_remote_filter_pull_through.py

```diff
@@ -0,0 +1,74 @@
+import subprocess
+import pytest
+import re
+
+from pulp_container.tests.functional.constants import (
+    REGISTRY_V2,
+    PULP_HELLO_WORLD_REPO,
+    PULP_FIXTURE_1,
+)
+
+
+@pytest.fixture
+def pull_and_verify(
+    capfd,
+    delete_orphans_pre,
+    local_registry,
+    registry_client,
+):
+    def _pull_and_verify(images, pull_through_distribution, includes, excludes):
+        distr = pull_through_distribution(includes, excludes)
+        for _, image_path in enumerate(images, start=1):
+            remote_image_path = f"{REGISTRY_V2}/{image_path}"
+            local_image_path = f"{distr.base_path}/{image_path}"
+
+            if excludes and re.match(".*fixture.*", image_path):
+                with pytest.raises(subprocess.CalledProcessError):
+                    local_registry.pull(local_image_path)
+                assert (
+                    re.search(
+                        ".*Repository not found.*",
+                        capfd.readouterr().err,
+                    )
+                    is not None
+                )
+                continue
```

Comment:
It is very hard to read and understand this part. We can pass expected values into this function and do the following:
```
if image_path not in XXX:
    with pytest.raises(subprocess.CalledProcessError) as exc_info:
        local_registry.pull(local_image_path)
    assert "Repository not found" in exc_info.value  # if possible, https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions
else:
    local_registry.pull(local_image_path)
```

Reply:
I tried to use the exception output, but the outputs from subprocesses seem to not be captured:
https://docs.pytest.org/en/6.2.x/capture.html

> If you want to capture on filedescriptor level you can use the capfd fixture which offers the exact same interface but allows to also capture output from libraries or subprocesses that directly write to operating system level output streams (FD1 and FD2).